res/schema.xml

```diff
@@ -458,4 +458,37 @@ METADATA
       );
     </sql>
   </revision>
+  <revision version="30" min_compatible="3">
+    <description>
+      Add tables to support nested crates.
+    </description>
+    <sql>
+      CREATE TABLE IF NOT EXISTS crateClosure (
+        parentId INTEGER,
+        childId INTEGER,
+        depth INTEGER
+      );
+
+      CREATE TABLE IF NOT EXISTS cratePath (
+        crateId INTEGER,
+        idPath TEXT,
+        namePath TEXT
```

Comment:
"This contains a / separated string of crate names"

Reply:
By the way: how do you treat names which includes a "/"?

Reply:
is it smart to store the namePath in the Database? This is a redundant information which can be wrong. 
How is guaranteed that this is always correct?

Reply:
Storing redundant information for optimizing queries is not an issue as long as you are able to guarantee consistency, either immediately or at least eventually. The method of choice for immediate consistency is encapsulation with hooks that are triggered when needed and everything enclosed in a transaction scope.

I remember that the my last review revealed that consistency cannot be guaranteed by the current implementation, because the transaction scoping was missing. Considering the fact that the whole hierarchy is abandoned and flattened when any inconsistencies are detected, this is a no-go for production and needs to be fixed.